DEVELOPMENT.md

```diff
@@ -269,3 +269,125 @@ If you need to add a new CRD type, you will need to add:
+// errs is an instance of *apis.FieldError, a common type in our validation code
+errs = errs.Also(ValidateEnabledAPIFields(ctx, "your feature name", requiredVersion))
+```
+
+If the user's cluster isn't configured with the required feature gate it'll
+return an error like this:
+
+```
+<your feature> requires "enable-api-fields" feature gate to be "alpha" but it is "stable"
+```
+
+### Unit Testing with Feature Gates
+
+Any new code you write that uses the `ctx` context variable is trivially
+unit tested with different feature gate settings. You should make sure
+to unit test your code both with and without a feature gate enabled to
+make sure it's properly guarded. See the following for an example of a
+unit test that sets the feature gate to test behaviour:
+
+```go
+featureFlags, err := config.NewFeatureFlagsFromMap(map[string]string{
+        "enable-api-fields": "alpha",
+})
+if err != nil {
+	t.Fatalf("unexpected error initializing feature flags: %v", err)
+}
+cfg := &config.Config{
+        FeatureFlags: featureFlags,
+}
+ctx := config.ToContext(context.Background(), cfg)
+if err := ts.TestThing(ctx); err != nil {
+	t.Errorf("unexpected error with alpha feature gate enabled: %v", err)
+}
+```
+
+### Example YAMLs
+<!-- TODO: This will be very different for Triggers -->
+<!-- Writing new YAML examples that require a feature gate to be set is easy.
+New YAML example files typically go in a directory called something like
+`examples/v1beta1/taskruns` in the root of the repo. To create a YAML that
+should only be exercised when the `enable-api-fields` flag is `alpha` just
+put it in an `alpha` subdirectory so the structure looks like:
+
+```
+examples/v1beta1/taskruns/alpha/your-example.yaml
+```
+
+This should work for both taskruns and pipelineruns.
+
+**Note**: To execute alpha examples with the integration test runner you
+must manually set the `enable-api-fields` feature flag to `alpha` in your
+testing cluster before kicking off the tests.
+
+When you set this flag to `stable` in your cluster it will prevent
+`alpha` examples from being created by the test runner. When you set
+the flag to `alpha` all examples are run, since we want to exercise
+backwards-compatibility of the examples under alpha conditions. -->
+
+### Integration Tests
+
+For integration tests we provide the [`requireGate` function](test/gate.go) which
+should be passed to the `setup` function used by tests:
+
+```go
+c, namespace := setup(ctx, t, requireGate("enable-api-fields", "alpha"))
+```
+
+This will Skip your integration test if the feature gate is not set to `alpha`
+with a clear message explaining why it was skipped.
+
+**Note**: As with running example YAMLs you have to manually set the `enable-api-fields`
+flag to `alpha` in your test cluster to see your alpha integration tests
+run. When the flag in your cluster is `alpha` _all_ integration tests are executed,
+both `stable` and `alpha`. Setting the feature flag to `stable` will exclude `alpha`
```

Comment:
I didn't get this. Why are both `stable` and `alpha` tests are executed when cluster is in `alpha`?